src/data/repositories/MasterLogFrameD2Repository.ts

```diff
@@ -0,0 +1,169 @@
+                ? d2Group.indicators
+                : [];
+
+        const description = this.isDataElementGroup(d2Group)
+            ? d2Group.displayDescription
+            : d2Group.description;
+
+        return {
+            id: d2Group.id,
+            name: description || d2Group.displayName,
+            type: indicatorType,
+            indicators: _(items)
+                .map(d2DataElement => ({ id: d2DataElement.id }))
+                .sortBy(indicator => indicator.id)
+                .value(),
+        };
+    }
+
+    private getDataElementGroupSet(codes: string[]): FutureData<Id[]> {
+        return apiToFuture(
+            this.api.models.dataElementGroupSets.get({
+                fields: { dataElementGroups: { id: true, code: true, name: true } },
+                paging: false,
+                filter: { code: { eq: this.MLF_CODE } },
+            })
+        ).map(response => {
+            const d2GroupSet = response.objects[0];
+            return d2GroupSet
+                ? this.filterByCodeAndExtractIds(codes, d2GroupSet.dataElementGroups)
+                : [];
+        });
+    }
+
+    private getIndicatorGroupSet(codes: string[]): FutureData<Id[]> {
+        return apiToFuture(
+            this.api.models.indicatorGroupSets.get({
+                fields: { id: true, indicatorGroups: { id: true, code: true, name: true } },
+                paging: false,
+                filter: { code: { eq: this.MLF_CODE } },
+            })
+        ).map(response => {
+            const d2GroupSet = response.objects[0];
+            return d2GroupSet
+                ? this.filterByCodeAndExtractIds(codes, d2GroupSet.indicatorGroups)
+                : [];
+        });
+    }
+
+    private filterByCodeAndExtractIds(
+        codes: string[],
+        groups: Array<{ id: Id; code: string; name: string }>
+    ): Id[] {
+        return groups
+            .filter(group => codes.includes(extractRegionCode(group.code ?? group.name)))
+            .map(group => group.id);
+    }
+
+    private isDataElementGroup(
+        group: D2DataElementGroup | D2IndicatorGroup
+    ): group is D2DataElementGroup {
+        return "dataElements" in group;
+    }
+
+    private isIndicatorGroup(
+        group: D2DataElementGroup | D2IndicatorGroup
+    ): group is D2IndicatorGroup {
+        return "indicators" in group;
+    }
+}
+
+const dataElementGroupFields = {
+    id: true,
+    displayName: true,
+    code: true,
+    dataElements: true,
+    displayDescription: true,
+} as const;
+
+const indicatorGroupFields = {
+    id: true,
+    displayName: true,
+    code: true,
+    indicators: true,
+    description: true,
+};
+
+type D2DataElementGroup = MetadataPick<{
+    dataElementGroups: { fields: typeof dataElementGroupFields };
+}>["dataElementGroups"][number];
+
+type D2IndicatorGroup = MetadataPick<{
+    indicatorGroups: { fields: typeof indicatorGroupFields };
+}>["indicatorGroups"][number];
```

Comment:
[note] As it's normal, we use this pattern throughout the code-base. As it's not terribly nice to read, we can create an abstraction (types/d2-api.ts) that captures this typical use case: 

`type D2IndicatorGroup = D2RecordType<"indicatorGroups", typeof indicatorGroupFields>` 

(D2RecordType is a proposal, it's a hard to find a name).